src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,263 @@
+			break;
+	}
+
+	return value;
+}
+
+xdebug_trace_flamegraph_context *xdebug_trace_flamegraph_init(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context;
+	char *used_fname;
+
+	tmp_flamegraph_context = xdmalloc(sizeof(xdebug_trace_flamegraph_context));
+	tmp_flamegraph_context->trace_file = xdebug_trace_open_file(fname, script_filename, options, (char**) &used_fname);
+	tmp_flamegraph_context->trace_filename = used_fname;
+
+	return tmp_flamegraph_context->trace_file ? tmp_flamegraph_context : NULL;
+}
+
+void *xdebug_trace_flamegraph_init_cost(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST;
+		tmp_flamegraph_context->stack = fg_stack_create();
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void *xdebug_trace_flamegraph_init_mem(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM;
+		tmp_flamegraph_context->stack = fg_stack_create();
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void xdebug_trace_flamegraph_deinit(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	if (NULL != context->stack) {
+		fg_stack_free(context->stack);
+	}
+
+	fclose(context->trace_file);
+	context->trace_file = NULL;
+	xdfree(context->trace_filename);
+
+	xdfree(context);
+}
+
+void xdebug_trace_flamegraph_write_header(void *ctxt)
+{
+}
+
+void xdebug_trace_flamegraph_write_footer(void *ctxt)
+{
+}
+
+char *xdebug_trace_flamegraph_get_filename(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	return context->trace_filename;
+}
+
+void xdebug_trace_flamegraph_function_entry(void *ctxt, function_stack_entry *fse, int function_nr)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+	flamegraph_stack_item *head = fg_stack_get_head(context->stack);
+	char *tmp_name, *prefix;
+
+	tmp_name = xdebug_show_fname(fse->function, 0, 0);
+
+	/* Insert an item into the custom stack which yields path prefix and
+	   a 0 value for the current function, if this function calls children
+	   they will append their costs to it. Once we reach out this function
+	   we will remove the element from the stack. */
+	if (NULL == head) {
+		/* Root function. */
+		fg_stack_push(context->stack, tmp_name);
+	} else {
+		prefix = xdmalloc(strlen(head->prefix) + 1 + strlen(tmp_name));
+		sprintf(prefix, "%s;%s", head->prefix, tmp_name);
+		fg_stack_push(context->stack, prefix);
+		xdfree(prefix);
```

Comment:
I was unsure about this as well, it duplicates the name because it can either be an owned string (the tmp_name variable) which is freed at the end of the enter hook, or a non-owned string (the xdebug stack item function name). But I probably can do better than this.